### PR TITLE
Initial support for pod eviction based on disk

### PR DESCRIPTION
**Pull request by derekwaynecarr**
This PR adds the following:
1. node reports disk pressure condition based on configured thresholds
2. scheduler does not place pods on nodes reporting disk pressure
3. kubelet will not admit any pod when it reports disk pressure
4. kubelet ranks pods for eviction when low on disk
5. kubelet evicts greediest pod

Follow-on PRs will need to handle:
1. integrate with new image gc PR (https://github.com/kubernetes/kubernetes/pull/27199)
2. container gc policy should always run (will not be launched from eviction, tbd who does that)
   1. this means kill pod is fine for all eviction code paths since container gc will remove dead container
3. min reclaim support will just poll summary provider (derek will do follow-on)
4. need to know if imagefs is same device as rootfs from summary (derek follow-on)

/cc @vishh @kubernetes/sig-node

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
There is more TODO here, but if we want a branch to collaborate on @vishh thought I would push it here.

**Comment by derekwaynecarr** [comment]
Will push the change to report DiskPressure next.

File: docs/proposals/kubelet-eviction.md [hunk 1]
```diff
@@ -348,20 +348,20 @@ If `nodefs` is triggering evictions, `kubelet` will sort pods based on the usage
...
+resource pressure, `kubelet` will attempt to reclaim at least `minimum-reclaim` amount of resource.
 
-Following are the flags through which `minimum-thresholds` can be configured for each evictable resource:
+Following are the flags through which `minimum-reclaim` can be configured for each evictable resource:
 
-`--minimum-eviction-thresholds="memory.available=0Mi,nodefs.available=500Mi,imagefs.available=2Gi"`
+`--eviction-minimum-reclaim="memory.available=0Mi,nodefs.available=500Mi,imagefs.available=2Gi"`
 
-The default `minimum-eviction-threshold` is `0` for all resources.
+The default `eviction-miniumum-reclaim` is `0` for all resources.
```
Comment (vreon) [inline docs/proposals/kubelet-eviction.md]: Typo: `minimum`

**Comment by derekwaynecarr** [comment]
@vishh - follow-on PRs from either of us will be needed to complete this feature, but this provides a decent basis for us to collaborate against when master branch reopens for kube 1.4.

**Comment by dchen1107** [comment]
cc/ @ronnielai

**Comment by derekwaynecarr** [comment]
Related umbrella issue:
https://github.com/kubernetes/kubernetes/issues/25239

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue https://github.com/kubernetes/kubernetes/issues/28121

**Comment by derekwaynecarr** [comment]
@vishh can this get a review this week so we can iterate?

File: cmd/kubelet/app/options/options.go [hunk 2]
```diff
@@ -276,6 +276,7 @@ func (s *KubeletServer) AddFlags(fs *pflag.FlagSet) {
 	fs.StringVar(&s.EvictionSoft, "eviction-soft", s.EvictionSoft, "A set of eviction thresholds (e.g. memory.available<1.5Gi) that if met over a corresponding grace period would trigger a pod eviction.")
 	fs.StringVar(&s.EvictionSoftGracePeriod, "eviction-soft-grace-period", s.EvictionSoftGracePeriod, "A set of eviction grace periods (e.g. memory.available=1m30s) that correspond to how long a soft eviction threshold must hold before triggering a pod eviction.")
 	fs.DurationVar(&s.EvictionPressureTransitionPeriod.Duration, "eviction-pressure-transition-period", s.EvictionPressureTransitionPeriod.Duration, "Duration for which the kubelet has to wait before transitioning out of an eviction pressure condition.")
+	fs.StringVar(&s.EvictionMinimumReclaim, "eviction-minimum-reclaim", s.EvictionMinimumReclaim, "A set of minimum reclaims (e.g. imagefs.available=2Gi) that describes the minimum amount of resource the kubelet will reclaim when eviction thresholds are met.")
```
Comment (ronnielai) [inline cmd/kubelet/app/options/options.go]: Could you be more specific about the eviction thresholds in "....when eviction thresholds are met"? I think it means eviction-soft thresholds. Just want to confirm.

File: plugin/pkg/scheduler/algorithm/predicates/predicates_test.go [hunk 3]
```diff
@@ -2807,3 +2807,73 @@ func TestPodSchedulesOnNodeWithMemoryPressureCondition(t *testing.T) {
...
+
+	for _, test := range tests {
+		fits, err := CheckNodeDiskPressurePredicate(test.pod, test.nodeInfo)
+		if fits != test.fits {
+			t.Errorf("%s: expected %v got %v", test.name, test.fits, fits)
+		}
+
+		if err != nil && err != ErrNodeUnderDiskPressure {
+			t.Errorf("%s: unexpected error: %v", test.name, err)
+			continue
```
Comment (ronnielai) [inline plugin/pkg/scheduler/algorithm/predicates/predicates_test.go]: The continue statement is not needed.

File: pkg/kubelet/eviction/eviction_manager.go [hunk 4]
```diff
@@ -67,6 +69,10 @@ func NewManager(
 	recorder record.EventRecorder,
 	nodeRef *api.ObjectReference,
 	clock util.Clock) (Manager, lifecycle.PodAdmitHandler, error) {
+
+	// TODO: pass this into the manager
+	withImageFs := false
```
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager.go]: Could you put with ImageFs inside the managerImpl struct? Then in the subsequent PR you can pass in the boolean to NewManager() to set it up properly.

File: pkg/kubelet/eviction/helpers.go [hunk 5]
```diff
@@ -186,6 +204,39 @@ func parseGracePeriods(expr string) (map[Signal]time.Duration, error) {
 	return results, nil
 }
 
+// parseMinimumReclaims parses the minimum reclaim statements
+func parseMinimumReclaims(expr string) (map[Signal]resource.Quantity, error) {
```
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Could you split parseMinimumReclaims into a separate commit?

File: pkg/kubelet/eviction/helpers.go [hunk 6]
```diff
@@ -186,6 +204,39 @@ func parseGracePeriods(expr string) (map[Signal]time.Duration, error) {
...
+		quantity, err := resource.ParseQuantity(parts[1])
+		if quantity.Sign() < 0 {
+			return nil, fmt.Errorf("negative eviction minimum reclaim specified for %v", signal)
+		}
+		if err != nil {
+			return nil, err
+		}
+		results[signal] = quantity
+	}
+	fmt.Printf("MIN RECLAIM: %s, RESULT: %v\n", expr, results)
```
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Please remove fmt.Printf....

File: pkg/kubelet/kubelet.go [hunk 7]
```diff
@@ -3252,6 +3252,64 @@ func (kl *Kubelet) setNodeMemoryPressureCondition(node *api.Node) {
...
+	// Note: The conditions below take care of the case when a new NodeDiskressure condition is
+	// created and as well as the case when the condition already exists. When a new condition
+	// is created its status is set to api.ConditionUnknown which matches either
+	// condition.Status != api.ConditionTrue or
+	// condition.Status != api.ConditionFalse in the conditions below depending on whether
+	// the kubelet is under disk pressure or not.
+	if kl.evictionManager.IsUnderDiskPressure() {
+		if condition.Status != api.ConditionTrue {
+			condition.Status = api.ConditionTrue
+			condition.Reason = "KubeletHasInsufficientDisk"
```
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: The wording here is colliding with the wording for OutOfDisk. I think the wording should indicate that the node is under disk pressure.

File: pkg/kubelet/kubelet.go [hunk 8]
```diff
@@ -3252,6 +3252,64 @@ func (kl *Kubelet) setNodeMemoryPressureCondition(node *api.Node) {
...
+			condition.Message = "kubelet has insufficient disk space available"
+			condition.LastTransitionTime = currentTime
+			kl.recordNodeStatusEvent(api.EventTypeNormal, "NodeHasInsufficientDisk")
```
Comment (ronnielai) [inline pkg/kubelet/kubelet.go]: The recorded event seems incorrect.

**Comment by derekwaynecarr** [comment]
Just got back from vacation, will update this PR tomorrow.  Thanks for review comments.

File: pkg/kubelet/kubelet.go [hunk 9]
```diff
@@ -3252,6 +3252,64 @@ func (kl *Kubelet) setNodeMemoryPressureCondition(node *api.Node) {
...
+	// created and as well as the case when the condition already exists. When a new condition
+	// is created its status is set to api.ConditionUnknown which matches either
+	// condition.Status != api.ConditionTrue or
+	// condition.Status != api.ConditionFalse in the conditions below depending on whether
+	// the kubelet is under disk pressure or not.
+	if kl.evictionManager.IsUnderDiskPressure() {
+		if condition.Status != api.ConditionTrue {
+			condition.Status = api.ConditionTrue
+			condition.Reason = "KubeletHasInsufficientDisk"
+			condition.Message = "kubelet has insufficient disk space available"
```
Comment (vishh) [inline pkg/kubelet/kubelet.go]: Is it worth including the availability here?

File: pkg/kubelet/eviction/helpers.go [hunk 10]
```diff
@@ -204,14 +255,52 @@ func memoryUsage(memStats *statsapi.MemoryStats) *resource.Quantity {
 	return resource.NewQuantity(usage, resource.BinarySI)
 }
 
-// podUsage aggregates usage of compute resources.
-// it supports the following memory and disk.
-func podUsage(podStats statsapi.PodStats) (api.ResourceList, error) {
+// localVolumeNames returns the set of volumes for the pod that are local
```
Comment (vishh) [inline pkg/kubelet/eviction/helpers.go]: nit: mention that these volumes consume disk.

Ideally, I'd argue that we should let the metrics summary API provide more information about what resources a volume consumes. For example, a tmpfs based empty dir consumes memory, and a regular empty dir consumes local disk and an nfs volume consumes remote storage, etc. Hard coding the list here is fine, but is probably not the cleanest solution. WDYT?

**Comment by vishh** [comment]
I'm mostly fine with this PR. I'l let @ronnielai continue the review.

File: pkg/kubelet/eviction/helpers.go [hunk 11]
```diff
@@ -204,14 +255,52 @@ func memoryUsage(memStats *statsapi.MemoryStats) *resource.Quantity {
...
+			volume.ConfigMap != nil ||
+			volume.GitRepo != nil {
+			result = append(result, volume.Name)
+		}
+	}
+	return result
+}
+
+// podDiskUsage aggregates pod disk usage for the specified stats to measure.
+func podDiskUsage(podStats statsapi.PodStats, pod *api.Pod, statsToMeasure []fsStats) (api.ResourceList, error) {
```
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Just a heads up. When this is used in the subsequent PR for eviction. https://github.com/kubernetes/kubernetes/blob/186c0c7b14dfc9403b131b68b70da08227c71a04/pkg/kubelet/image_manager.go#L51 provides the method to delete all unused images.

File: pkg/kubelet/eviction/eviction_manager.go [hunk 12]
```diff
@@ -87,12 +94,17 @@ func (m *managerImpl) Admit(attrs *lifecycle.PodAdmitAttributes) lifecycle.PodAd
 	if len(m.nodeConditions) == 0 {
 		return lifecycle.PodAdmitResult{Admit: true}
 	}
-	notBestEffort := qos.BestEffort != qos.GetPodQOS(attrs.Pod)
-	if notBestEffort {
-		return lifecycle.PodAdmitResult{Admit: true}
+
```
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager.go]: Where do you plan to invoke image GC when there's disk pressure?(https://github.com/kubernetes/kubernetes/blob/186c0c7b14dfc9403b131b68b70da08227c71a04/pkg/kubelet/image_manager.go#L51 provides the method to delete all unused images.)

File: pkg/kubelet/eviction/eviction_manager_test.go [hunk 13]
```diff
@@ -49,7 +49,7 @@ func TestMemoryPressure(t *testing.T) {
 	podMaker := func(name string, requests api.ResourceList, limits api.ResourceList, memoryWorkingSet string) (*api.Pod, statsapi.PodStats) {
```
Comment (ronnielai) [inline pkg/kubelet/eviction/eviction_manager_test.go]: Please also add a test for disk pressure.

File: cmd/kubelet/app/options/options.go (hunk 2 above)
Comment (derekwaynecarr) [inline cmd/kubelet/app/options/options.go]: Its when any eviction threshold is met (soft or hard).

Comment (derekwaynecarr) [inline cmd/kubelet/app/options/options.go]: i will update the text to state that "describes the minimum amount of resource the kubelet will reclaim when performing a pod eviction if that resource is under pressure."

File: pkg/kubelet/eviction/eviction_manager.go (hunk 4 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager.go]: it's actually not needed on managerImpl once we build the ranking funcs.

I will leave it where it is now, and just pass it into NewManager in follow-up.

File: pkg/kubelet/kubelet.go (hunk 7 above)
Comment (derekwaynecarr) [inline pkg/kubelet/kubelet.go]: hmm... i see what you mean.

File: pkg/kubelet/eviction/eviction_manager.go (hunk 12 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager.go]: i was going to try and structure it as a series of actions, this call was not around when this PR was opened.  will think on the best way to structure it cleanly in a follow-on.

File: pkg/kubelet/eviction/eviction_manager_test.go (hunk 13 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager_test.go]: will do.

File: pkg/kubelet/eviction/helpers.go (hunk 10 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: i agree long term.  i will add a // TODO for follow-on.

for this PR, i will just ensure that empty dir is only included when medium is not memory.

File: pkg/kubelet/eviction/eviction_manager.go (hunk 4 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/eviction_manager.go]: actually, i fixed this up, it cant be passed in NewManager because knowing if imagefs and rootfs are unified requires knowledge that is runtime dependent.  i am going to build the rank functions in the Start() and just start the manager after we start cAdvisor.

**Comment by derekwaynecarr** [comment]
marking this do-not-merge while i tease out some parts of this PR that are a chore to rebase into smaller PRs that hopefully can merge quickly.

**Comment by derekwaynecarr** [comment]
@ronnielai - I think the prerequisite prs have all merged, please take a look.  once this goes in, will look at invoking imagegc

File: pkg/kubelet/eviction/helpers.go (hunk 10 above)
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Please add the TODO

File: pkg/kubelet/eviction/helpers.go [hunk 14]
```diff
@@ -248,14 +254,52 @@ func memoryUsage(memStats *statsapi.MemoryStats) *resource.Quantity {
...
+		if hasFsStats(statsToMeasure, fsStatsLogs) {
+			disk.Add(*diskUsage(container.Logs))
+		}
+	}
+	if hasFsStats(statsToMeasure, fsStatsLocalVolumeSource) {
+		volumeNames := localVolumeNames(pod)
+		for _, volumeName := range volumeNames {
+			for _, volumeStats := range podStats.VolumeStats {
+				if volumeStats.Name == volumeName {
+					disk.Add(*diskUsage(&volumeStats.FsStats))
```
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: I think you can break after disk.add in line 287.

File: pkg/kubelet/eviction/types.go [hunk 15]
```diff
@@ -30,6 +30,22 @@ type Signal string
...
 	// SignalMemoryAvailable is memory available (i.e. capacity - workingSet), in bytes.
 	SignalMemoryAvailable Signal = "memory.available"
+	// SignalNodeFsAvailable is amount of storage available on filesystem that kubelet uses for volumes, daemon logs, etc.
+	SignalNodeFsAvailable Signal = "nodefs.available"
+	// SignalImageFsAvailable is amount of storage available on filesystem that container runtime uses for for storing images and container writable layers.
+	SignalImageFsAvailable Signal = "imagefs.available"
+)
+
+// fsStats defines the types of filesystem stats to collect.
+type fsStats string
```
Comment (ronnielai) [inline pkg/kubelet/eviction/types.go]: fsStats kind of collide with FsStats. could you name it fsStatType?

File: pkg/kubelet/eviction/helpers.go [hunk 16]
```diff
@@ -608,3 +675,21 @@ func isSoftEviction(thresholds []Threshold, starvedResource api.ResourceName) bo
 	}
 	return true
 }
+
+// buildresourceToRankFunc returns ranking functions associated with resources
```
Comment (ronnielai) [inline pkg/kubelet/eviction/helpers.go]: Typo. should be buildResourceToRankFunc

File: pkg/kubelet/eviction/helpers.go (hunk 10 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: done

File: pkg/kubelet/eviction/helpers.go (hunk 14 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/helpers.go]: good point, fixed.

File: pkg/kubelet/eviction/types.go (hunk 15 above)
Comment (derekwaynecarr) [inline pkg/kubelet/eviction/types.go]: sure.

**Comment by derekwaynecarr** [comment]
@ronnielai -- all feedback addressed, lets get this in so we can continue to iterate.

**Comment by ronnielai** [comment]
LGTM. Please address the test failure.

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue #ignore

**Comment by pmorie** [comment]
@k8s-bot test this issue: #29703

**Comment by derekwaynecarr** [comment]
It appears something is wrong with collecting root container stats on GCE that I was not seeing locally.  Will investigate tomorrow.

**Comment by derekwaynecarr** [comment]
@ronnielai - tests passed, turns out cadvisor is slower to warm on gce, so moved diskinfoprovider call into a one time setup in each synchronize call.

**Comment by derekwaynecarr** [comment]
fixed rebase error, retagging.

**Comment by derekwaynecarr** [comment]
@k8s-bot test this issue #ignore
